# Table of Content
01. Import Libraries
02. Import Data
03. Finding Mixed-type Data
04. Data Accuracy
05. Finding Missing Values
06. Finding Duplicates
07. Export Data

# 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# Define the main project folder path
path = r'C:\Users\saich\Desktop\CareerFoundry\Data Immersion\Achievement 4 Python Fundamentals for Data Analysts\04-2023 Instacart Basket Analysis (github)'

In [3]:
# Import 'orders_wrangled' data set from 'Prepared Data' folder
ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [4]:
ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [5]:
ords.shape

(3421083, 6)

# 03. Finding Mixed-type Data

In [6]:
# Check for any mixed-type columns - no output means no mixed-type data
for col in ords.columns.tolist():
    weird = (ords[[col]].applymap(type) != ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (ords[weird]) > 0:
        print (col)

There is no mixed-type data in 'ords' dataframe. 

# 04. Data Accuracy

In [7]:
# Check the descriptive statistics for numerical data
ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

There is no unusual data in 'ords' dataframe.

# 05. Finding Missing Values

In [8]:
# Check for missing values
ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in 'days_since_prior_order' column.

In [9]:
# Take a look at the missing values
ords_nan = ords.loc[ords['days_since_prior_order'].isnull() == True]

In [10]:
ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [11]:
# Check the values of 'order_number' when 'days_since_prior_order' is missing
ords_nan['order_number'].value_counts(dropna = False)

1    206209
Name: order_number, dtype: int64

From the data check above, the 206209 missing values in 'days_since_prior_order' column are all come with an order number of 1. Since there can't be any days since prior order before the very first order, therefore no action is needed. 

# 06. Finding Duplicates

In [12]:
# Check for duplicates
ords.duplicated().sum()

0

In [13]:
ords.loc[ords.duplicated() == True]

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There is no duplicate data in 'ords' dataframe. 

# 07. Export Data

In [14]:
# Dimension check for 'ords' dataframe after data consistency check
ords.shape

(3421083, 6)

In [15]:
# Export 'ords' dataframe to 'Prepared Data' folder in csv format
ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))